File: View Controllers/OrganizationsViewController.m
//
//  OrganizationsViewController.m
//  TigerConnect Demo
//
//  Created by Oren Zitoun on 2/2/16.
//  Copyright © 2016 TigerText. All rights reserved.
//

#import <TTKit/TTKit.h>

#import "OrganizationsViewController.h"
#import "OrganizationCell.h"

@interface OrganizationsViewController () <UITableViewDataSource, UITableViewDelegate, NSFetchedResultsControllerDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic) NSFetchedResultsController *resultsController;

@end

@implementation OrganizationsViewController

// Handling close button press
- (IBAction)onCloseButton:(id)sender
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Loading the organization fetched results controller to display and monitor all organizations
    self.resultsController = [[TTKit sharedInstance] organizationsFetchControllerWithDelegate:self];
    [self.tableView registerNib:[UINib nibWithNibName:NSStringFromClass(OrganizationCell.class) bundle:nil] forCellReuseIdentifier:NSStringFromClass(OrganizationCell.class)];
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    [self.tableView reloadData];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.resultsController.fetchedObjects.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [OrganizationCell cellHeight];
}

- (void)configureCell:(OrganizationCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    TTOrganization *organization = [self.resultsController objectAtIndexPath:indexPath];
    cell.item = organization;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    OrganizationCell *cell = (OrganizationCell *)[tableView dequeueReusableCellWithIdentifier:NSStringFromClass(OrganizationCell.class)];
    [self configureCell:cell forRowAtIndexPath:indexPath];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    TTOrganization *organization = [self.resultsController objectAtIndexPath:indexPath];
    [[TTKit sharedInstance] setCurrentOrganization:organization];
    [self onCloseButton:nil];
}

#pragma mark - NSFetchedResultsControllerDelegate

// Refreshing the table view after any update to the organizations
- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller
{
    [self.tableView reloadData];
}

@end

File: TC Swift Demo/ViewControllers/OrganizationsViewController.swift
//
//  OrganizationsViewController.swift
//  TC Swift Demo
//
//  Created by Oren Zitoun on 2/12/16.
//  Copyright © 2016 TigerConnect. All rights reserved.
//

import Foundation

class OrganizationsViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, NSFetchedResultsControllerDelegate {
    
    @IBOutlet weak var tableView: UITableView!
    var resultsController: NSFetchedResultsController<NSFetchRequestResult>!
    
    @IBAction func onCloseButton(_ sender: AnyObject) {
        self.dismiss(animated: true, completion: nil)
    }
    
    let organizationsCellIdentifier = "OrganizationCell"

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.resultsController = TTKit.sharedInstance().organizationsFetchController(with: self)
        self.tableView.register(UINib(nibName: organizationsCellIdentifier, bundle: nil), forCellReuseIdentifier: organizationsCellIdentifier)
        self.tableView.tableFooterView = UIView.init(frame: CGRect.zero)
        self.tableView.reloadData()
    }
    
    func configureCell(_ cell: OrganizationCell, indexPath: IndexPath) {
        let organization = self.resultsController.object(at: indexPath)
        cell.item = organization as? TTOrganization
    }
    
    //MARK: UITableViewDataSource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let count = self.resultsController.fetchedObjects?.count
        return count!
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return OrganizationCell.cellHeight()
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: String(describing: organizationsCellIdentifier)) as! OrganizationCell
        self.configureCell(cell, indexPath: indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let organization = self.resultsController.object(at: indexPath) as? TTOrganization
        TTKit.sharedInstance().setCurrentOrganization(organization)
        self.onCloseButton(self)
    }
    
    //MARK: NSFetchedResultsControllerDelegate
    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
        self.tableView.reloadData()
    }
}
